stm32: adc: fix build of ADC channel setup of STM32L4 series

benediktibk opened the pull request:
Fixes #43750

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -852,6 +852,29 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
 	}
 
 #elif defined(CONFIG_SOC_SERIES_STM32L4X)
+#if defined(CONFIG_SOC_STM32L412XX) || \
+	defined(CONFIG_SOC_STM32L422XX) || \
+	defined(CONFIG_SOC_STM32L432XX) || \
+	defined(CONFIG_SOC_STM32L433XX) || \
+	defined(CONFIG_SOC_STM32L452XX) || \
+	defined(CONFIG_SOC_STM32L462XX) || \
+	defined(CONFIG_SOC_STM32L4S5XX) || \
+	defined(CONFIG_SOC_STM32L4R5XX) || \
+	defined(CONFIG_SOC_STM32L4R9XX)
```
What about `#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)` ?

benediktibk replied:
I would then go for something like?

    #if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)  && DT_NODE_HAS_STATUS(DT_NODELABEL(adc3), okay)
    ...
    #elif DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
    ...
    #endif

FRASTM replied:
Note that some stm32 mcus like stm32G4 or U5 have TEMP sensor on other ADC than 3: 
```
- The internal reference voltage (V REFINT ), connected to ADC1, 3, 4 and 5
- The internal temperature sensor (V TS ), connected to ADC1 and 5
```

benediktibk commented on drivers/adc/adc_stm32.c:
```diff
@@ -855,13 +855,24 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
 	/* Temperature available on ADCs 1 and 3.
 	 * Refint available on ADC 1 only.
 	 */
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && \
+	DT_NODE_HAS_STATUS(DT_NODELABEL(adc3), okay)
 	if ((adc == ADC1 || adc == ADC3) &&
 		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
 		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
 	} else if (adc == ADC1 &&
 		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
 		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
 	}
+#elif DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
+	if ((adc == ADC1) &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
+		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
+	} else if (adc == ADC1 &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
+		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
+	}
+#endif
```
Me neither. What about a macro function, e.g.

    #if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && \
	 DT_NODE_HAS_STATUS(DT_NODELABEL(adc3), okay)
    #define IS_ADC_WITH_TEMPERATURE(adc) (adc == ADC1 || adc == ADC3)
    #elif DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
    #define IS_ADC_WITH_TEMPERATURE(adc) (adc == ADC1)
    #else
    #define IS_ADC_WITH_TEMPERATURE(adc) (false)
    #endif

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -798,93 +798,77 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id) ||
-		(adc == ADC5 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC5) == channel_id)) {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && \
+	DT_PROP(DT_NODELABEL(adc1), has_temp_channel)
+	if (adc == ADC1 &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
```
From this point, you should be able to do a "mere" (modulo the series api adaptation):
```
	if (config->has_temp_channel &&
		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
```
This requires adding fields to device config struct, but likely worse the gain  considering this function.

erwango commented on dts/arm/st/l4/stm32l4r5.dtsi:
```diff
@@ -312,3 +312,8 @@
 		label = "OTGFS_PHY";
 	};
 };
+
+&adc1 {
+	has-temp-channel;
+	has-vref-channel;
+};
```
Please don't add phandle fixups in this file, but rather overwrite the nodes directly:
```
		adc1: adc@foo{
			has-temp-channel;
			has-vref-channel;
		};
	};
};
```

erwango reviewed (changes requested):
Thanks for this nice fix & addition.
I guess new property names could be bikeshed but given the overall improvement, let's go for it.
One request: please  split the dts/bindings from the driver bits into 2 separate commits.
Otherwise LGTM

benediktibk commented:
> Thanks for this nice fix & addition. I guess new property names could be bikeshed but given the overall improvement, let's go for it. One request: please split the dts/bindings from the driver bits into 2 separate commits. Otherwise LGTM

Thanks for guiding me towards this approach, looks way better now than previously.

erwango reviewed (commented):
This is nitpicky, but in general we recommend, for bisectibility and ease of the review:
- put the dts and bindings in same commit (some also split them a part) > Having dts and bindings in same commit eases the review, we can see that dts additions match bindings
- put the drivers bits in a dedicated following commit > This way driver is functional before and after the change (... in this case driver was broken but, yet...)

benediktibk committed:
dts: arm: stm32: introduce ADC properties for temp and vref channel

Introduce ADC properties which indicated if the ADC instances have
dedicated channels for the internal temperature sensor or voltage
reference.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
stm32: adc: use ADC properties to configure temp and vref channels

Use the devicetree properties to determine if the dedicated temperature or
voltage reference channels should be configured for the ADC.
Fixes #43750.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

gmarull reviewed (approved):
LGTM. I'll save this as an example of HAL not helping, but increasing code complexity.
